zbark/debark.go

```diff
@@ -0,0 +1,144 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zbark
+
+import (
+	"github.com/uber-go/zap"
+	"github.com/uber-go/zap/zwrap"
+
+	"github.com/uber-common/bark"
+)
+
+// Debarkify wraps bark.Logger to make it compatible with zap's JSON logger
+func Debarkify(b bark.Logger) zap.Logger {
+	if wrapper, ok := b.(*barker); ok {
+		return wrapper.zl
+	}
+	return &zapper{
+		lvl: zap.DebugLevel,
+		bl:  b,
+	}
+}
+
+type zapper struct {
+	lvl zap.Level
+	bl  bark.Logger
+}
+
+func (z *zapper) DFatal(msg string, fields ...zap.Field) {
+	// TODO: Implement development/DFatal
+	z.Error(msg, fields...)
+}
+
+func (z *zapper) Log(l zap.Level, msg string, fields ...zap.Field) {
+	if l < z.lvl {
+		return
+	}
+	switch l {
+	case zap.DebugLevel:
+		z.Debug(msg, fields...)
+	case zap.InfoLevel:
+		z.Info(msg, fields...)
+	case zap.WarnLevel:
+		z.Warn(msg, fields...)
+	case zap.ErrorLevel:
+		z.Error(msg, fields...)
+	case zap.PanicLevel:
+		z.Panic(msg, fields...)
+	case zap.FatalLevel:
+		z.Fatal(msg, fields...)
+	default:
+		panic("passed an unknown zap.Level")
+	}
+}
+
+func (z *zapper) Level() zap.Level {
+	return z.lvl
+}
+
+// Change the level of this logger, as well as all its ancestors and
+// descendants. This makes it easy to change the log level at runtime
+// without restarting your application.
+func (z *zapper) SetLevel(l zap.Level) {
+	z.lvl = l
+}
+
+// Create a child logger, and optionally add some context to that logger.
+func (z *zapper) With(fields ...zap.Field) zap.Logger {
+	return &zapper{
+		lvl: z.lvl,
+		bl:  z.bl.WithFields(zapToBark(fields)),
+	}
+}
+
+func (z *zapper) StubTime() {
+	// StubTime stops the logger from including the current time in each
+	// message. Instead, it always reports the time as Unix epoch 0. (This is
+	// useful in tests and examples.)
+	panic("fake implementation")
+}
+
+func (z *zapper) Check(l zap.Level, msg string) *zap.CheckedMessage {
+	if l < z.lvl {
+		return nil
+	}
+	return zap.NewCheckedMessage(z, l, msg)
+}
+
+type zapperBarkFields struct {
+	zwrap.KeyValueMap
+}
+
+func (zbf zapperBarkFields) Fields() map[string]interface{} {
+	return zbf.KeyValueMap
+}
+
+func zapToBark(zfs []zap.Field) bark.LogFields {
+	zbf := zapperBarkFields{make(zwrap.KeyValueMap, len(zfs))}
+	for _, zf := range zfs {
+		_ = zf.AddTo(zbf)
```

Comment:
it was a note to myself to consider checking the error value and do something if possible.